pp.c

```diff
@@ -4882,22 +4882,86 @@ PP(pp_fc)
 
 PP(pp_trim)
 {
-    dTARGET;
     dSP;
+    dTARGET;
     SV *source = TOPs;
+    STRLEN len;
+    const U8 *start;
     SV *dest;
 
     SvGETMAGIC(source);
 
+    if(SvOK(source))
+        start = (const U8*)SvPV_nomg_const(source, len);
+    else {
+        if(ckWARN(WARN_UNINITIALIZED))
+            report_uninit(source);
+        start = (const U8*)"";
+        len = 0;
+    }
+
+    if(DO_UTF8(source)) {
+        const U8 *end = start + len;
+        const U8 *s;
+
+        while(len) {
+            STRLEN thislen;
+            if(!isSPACE_utf8_safe(start, end))
+                break;
+            start += (thislen = UTF8SKIP(start));
+            len -= thislen;
+        }
+
+        /* We can't walk backwards from the end; instead we must walk forwards
+         * from here remembering the final non-space position we saw */
```

Comment:
Why can't we walk backwards from the end?

UTF-8 allows it, and we have an inline function (utf8_hop_back()) to walk backwards through a UTF-8 string.

Reply:
Agreed.